core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -232,117 +267,167 @@ public boolean contains(IracTombstoneInfo tombstone) {
+         return Action.REMOVE_TOMBSTONE;
+      } else if (!info.isPrimary()) {
+         // backup owner, notify primary owner to check if the tombstone can be cleanup
+         return iracManager.running().containsKey(tombstone.getKey()) ? Action.KEEP_TOMBSTONE : Action.NOTIFY_PRIMARY_OWNER;
+      } else {
+         // primary owner, check all remote sites.
+         return iracManager.running().containsKey(tombstone.getKey()) ? Action.KEEP_TOMBSTONE : Action.CHECK_REMOTE_SITE;
+      }
+   }
+
+   private final class CleanupTask implements Function<Void, CompletionStage<Void>>, Runnable {
+      private final Collection<IracTombstoneInfo> tombstoneToCheck;
+      private final IntSet tombstoneToKeep;
+      private final int id;
+      private volatile boolean failedToCheck;
+
+      private CleanupTask(Collection<IracTombstoneInfo> tombstoneToCheck) {
+         this.tombstoneToCheck = tombstoneToCheck;
+         tombstoneToKeep = IntSets.concurrentSet(tombstoneToCheck.size());
+         failedToCheck = false;
+         id = tombstoneToCheck.hashCode();
       }
 
-      CompletionStage<Void> checkRemoteSites() {
+      CompletableSource check() {
+         if (log.isTraceEnabled()) {
+            log.tracef("[cleanup-task-%d] Running cleanup task with %s tombstones to check", id, tombstoneToCheck.size());
+         }
+         if (tombstoneToCheck.isEmpty()) {
+            return Completable.complete();
+         }
+         List<Object> keys = tombstoneToCheck.stream()
+               .map(IracTombstoneInfo::getKey)
+               .collect(Collectors.toList());
+         IracTombstoneRemoteSiteCheckCommand cmd = commandsFactory.buildIracTombstoneRemoteSiteCheckCommand(keys);
          // if one of the site return true (i.e. the key is in updateKeys map, then do not remove it)
-         AggregateCompletionStage<Boolean> stage = CompletionStages.orBooleanAggregateCompletionStage();
+         AggregateCompletionStage<Void> stage = CompletionStages.aggregateCompletionStage();
          for (XSiteBackup backup : asyncBackups) {
             if (takeOfflineManager.getSiteState(backup.getSiteName()) == SiteState.OFFLINE) {
                continue; // backup is offline
             }
             // we don't need the tombstone to query the remote site
-            IracTombstoneRemoteSiteCheckCommand cmd = commandsFactory.buildIracTombstoneRemoteSiteCheckCommand(tombstone.getKey());
-            stage.dependsOn(rpcManager.invokeXSite(backup, cmd));
+            stage.dependsOn(rpcManager.invokeXSite(backup, cmd).thenAccept(this::mergeIntSet));
          }
          // in case of exception, keep the tombstone
-         return stage.freeze()
-               .exceptionally(toTrueFunction())
-               .thenComposeAsync(this, blockingExecutor);
+         return Completable.fromCompletionStage(
+               stage.freeze()
+                     .exceptionally(this::onException)
+                     .thenComposeAsync(this, blockingExecutor));
```

Comment:
well, I'm not sure if it is a good idea. This will block a "blocking-thread" until the cross-site request is received, but the thread could be doing something else.

Reply:
Not sure I understand. If the `apply` method doesn't block there is no reason to use the `blockingExecutor` in the first place. Are you thinking this operation is going to be long running but not blocking and could cause latency issues if invoked on an event loop thread?

Reply:
The `Flowable` will run in a "blocking-thread" and the `BlockingManager#thenApplyBlocking` will block the `Flowable` iteration.

Reply:
> The `Flowable` will run in a "blocking-thread" and the `BlockingManager#thenApplyBlocking` will block the `Flowable` iteration.

This isn't how `BlockingManager.thenApplyBlocking` works. It will only run the apply method in a blocking thread, which from the code seems to be what you are trying to do.  The problem is the returned stage from `thenComposeAsync` will continue processing on the blocking thread, which we **don't** want and is the sole purpose of `BlockingManger`.

The question is _what_ needs to run on a blocking thread?

Reply:
https://github.com/infinispan/infinispan/blob/main/core/src/main/java/org/infinispan/util/concurrent/BlockingManagerImpl.java#L179-L190
^ it looks like it waits for the `CompletionStage` before invoking the `Function`.

Reply:
Only if the current thread is blocking. And in that case the provided stage should be already completed in pretty much every case, that is the point of the class.

But I agree to remove ambiguity, we should probably add a check if the stage is completed or not to the first if.

Reply:
This will be invoked from a JGroups thread on a response though, so it will never be done on a blocking there here afai can see.

Reply:
The problem with the code it is currently is that the apply will be ran in a blocking thread which will in turn complete the returned stage in a blocking thread, which means we will leak the blocking thread out to callers, which we don't want. The only time we allow returning the completed stage on a blocking thread is if the current thread is a blocking thread. Some vary specific edge cases, which tbh might be removed sometime soon will even allow for this occur in the first place.

But as I mentioned _what_ actually needs to be ran on a blocking thread? Looking at `CleanupTask#apply` it isn't apparent what requires a blocking thread there.